snuba/querylog/factory.py

```diff
@@ -0,0 +1,34 @@
+from snuba.clickhouse.query_dsl.accessors import get_time_range
+from snuba.datasets.dataset import Dataset
+from snuba.datasets.entities.factory import get_entity
+from snuba.datasets.factory import get_dataset_name
+from snuba.query.data_source.projects_finder import ProjectsFinder
+from snuba.query.logical import Query
+from snuba.querylog.query_metadata import SnubaQueryMetadata
+from snuba.request import Request
+from snuba.utils.metrics.timer import Timer
+
+
+def create_snuba_query_metadata(
```

Comment:
Why not just make the SnubaQueryMetadata constructor generate these things? What purpose does the factory function serve?

Reply:
This seemed like a good option, I just explored the idea of using default values in the constructor for SnubaQueryMetadata to simulate constructor override. And just generate them like this if they are None.

It turns out that this doesn't seem viable to implement given that the class is @dataclass, I can't override __init__ and it ends up messing up typing in the whole class.

I dont really know enough about @dataclass to know if its safe to remove it and just define my own constructor.

Reply:
it is safe to remove and provide your own constructor